### PR TITLE
feat(gradle): upgrade gradle to 6.4.1

### PR DESCRIPTION
**Pull request by DarkWeird**
Testing:
  1. Standalone:
        * run `gradle jar`
  2. As part of terasology:
       * clone it in Terasology/libs
       * run `gradle game` on Terasology dir

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -22,23 +28,32 @@ targetCompatibility = 1.8
...
-        url "http://artifactory.terasology.org/artifactory/virtual-repo-live"
+        def repoViaEnv = System.getenv()["RESOLUTION_REPO"]
+        if (rootProject.hasProperty("alternativeResolutionRepo")) {
+            // If the user supplies an alternative repo via gradle.properties then use that
+            name "from alternativeResolutionRepo property"
+            url alternativeResolutionRepo
+        } else if (repoViaEnv != null && repoViaEnv != "") {
+            name "from \$RESOLUTION_REPO"
+            url = repoViaEnv
+        } else {
```
Comment (skaldarnar) [inline build.gradle:39]: Why is this necessary, or what is it good for?

File: build.gradle [hunk 2]
```diff
@@ -22,23 +28,32 @@ targetCompatibility = 1.8
...
-}
-
 configurations {
     codeMetrics
 }
 
 dependencies {
 
-    codeMetrics(group: 'org.terasology.config', name: 'codemetrics', version: '1.0.0', ext: 'zip')
+    codeMetrics(group: 'org.terasology.config', name: 'codemetrics', version: '1.3.2', ext: 'zip')
```
Comment (skaldarnar) [inline build.gradle:54]: https://github.com/MovingBlocks/TeraConfig was recently released as v1.5.0, containing a fix for the license header regex. If we're doing an upgrade here, let's go for 1.5.0 😉

File: build.gradle [hunk 3]
```diff
@@ -87,37 +102,69 @@ pmd {
...
 
-findbugs {
+spotbugs {
     ignoreFailures = true
     effort = 'max'
     reportLevel = 'medium'
-    toolVersion = '3.0.1'
-    excludeFilterConfig = resources.text.fromArchiveEntry(configurations.codeMetrics, "findbugs/findbugs-exclude.xml")
+    toolVersion = '4.0.0'
+    excludeFilter = new File(rootDir, "config/metrics/findbugs/findbugs-exclude.xml")
```
Comment (skaldarnar) [inline build.gradle:110]: Is it safe to just assume the config is extracted to `config/metrics`? And should we also rename that folder inside to `spotbugs`?

File: build.gradle [hunk 4]
```diff
@@ -1,14 +1,20 @@
+buildscript {
+    dependencies {
+        classpath "gradle.plugin.com.github.spotbugs.snom:spotbugs-gradle-plugin:4.0.0"
+    }
+}
+
 plugins {
     id "com.jfrog.artifactory" version "3.1.1"
+    id "java-library"
```
Comment (keturn) [inline build.gradle:9]: might as well convert all the `apply plugin` to use `plugins {}` while we're at it

File: build.gradle [hunk 5]
```diff
@@ -1,14 +1,20 @@
+buildscript {
+    dependencies {
+        classpath "gradle.plugin.com.github.spotbugs.snom:spotbugs-gradle-plugin:4.0.0"
```
Comment (keturn) [inline build.gradle:3]: suggest using plugins block instead: https://plugins.gradle.org/plugin/com.github.spotbugs

_unless_ you are specifically trying to keep the PRs for upgrading gradle and upgrading spotbugs separate.

File: build.gradle (hunk 1 above)
Comment (keturn) [reply build.gradle:39]: I recognize this code, along with the `publishing` stuff below, from Terasology's publish.gradle.

It's the stuff I'm always afraid of breaking because it's not part of a workflow I use so I don't know how to test it or how much of it is covered by CI.

File: gradle/wrapper/gradle-wrapper.properties [hunk 6]
```diff
@@ -1,6 +1,6 @@
-#Mon Jul 27 16:18:33 CEST 2015
+#Fri Jul 10 10:40:08 MSK 2020
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.5-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.4.1-bin.zip
```
Comment (keturn) [inline gradle/wrapper/gradle-wrapper.properties:6]: If you've loaded this in IntelliJ, it may have asked for the source distribution so it can do completions better. That's done by changing this `bin` to an `all` or by doing some [configuration to the wrapper task](https://docs.gradle.org/6.4.1/userguide/gradle_wrapper.html#customizing_wrapper).

File: gradlew.bat [hunk 7]
```diff
@@ -1,90 +1,90 @@
...
-rem Set variable GRADLE_EXIT_CONSOLE if you need the _script_ return code instead of
-rem the _cmd.exe /c_ return code!
-if  not "" == "%GRADLE_EXIT_CONSOLE%" exit 1
-exit /b 1
-
-:mainEnd
-if "%OS%"=="Windows_NT" endlocal
-
-:omega
+@if "%DEBUG%" == "" @echo off
```
Comment (keturn) [inline gradlew.bat:1]: GitHub says every line here changed, which probably means DOS/Unix line endings. I assume `.bat` files need DOS endings, but I don't how how to tell from GitHub's web view which this is currently.

I don't know if the default git settings know about `.bat`, or if we have config in the other repos for this.

Semi-relatedly:

If this is upgrading all the way from 2.x, I expect there are new versions of both `gradlew.sh` and `gradlew.bat`. Those changes over the years have been pretty minor but they do catch some edge cases, so it's good to include them when we do upgrades.

File: build.gradle (hunk 1 above)
Comment (Cervator) [reply build.gradle:39]: This is the scaffolding that lets the extra magic in `gradle.properties` work for Terasology, yep. It lets you do the stuff like:

```
alternativeResolutionRepo=http://artifactory.terasology.org/artifactory/virtual-nanoware-and-remote
publishRepo=nanoware-snapshot-local`
```

And you'll be resolving + publishing artifacts against the Nanoware repos in Artifactory, excluding anything that's _live_ in the regular Terasology repos.

Jenkins uses the same approach for Nanoware fork builds, they all get those values via env vars added on the [Nanoware _folder_ in Jenkins](http://jenkins.terasology.io/teraorg/job/Nanoware/configure), without needing it in a `gradle.properties` at all.

It is kinda overkill in here, but did let me verify via the Nanoware snapshot repo that a publish from this project looked right.

It is in essence entirely outside the normal functionality, unless those props or env vars are set it does nothing and goes with the defaults.

File: build.gradle (hunk 3 above)
Comment (Cervator) [reply build.gradle:110]: Not really safe. I made & published binaries from my local workspace without realizing this had been hooked up, it just didn't crash since I didn't trigger the analytics, heh. In the main Terasology engine the extract task is tied to some stuff to make sure it runs, we should do that here as well.

File: build.gradle (hunk 2 above)
Comment (Cervator) [reply build.gradle:54]: Did this in some minor changes I'll add to a new branch 👍

File: build.gradle (hunk 4 above)
Comment (Cervator) [reply build.gradle:9]: Sadly I think last I looked at this it didn't work for all plugins - a few older ones or legacy style ones aren't compatible. Something-something need the old school `buildscript` dependency thing to work or so. I tried it _real_ quick - and it does seem like the spotbugs plugin works the new way now. But the config / codeMetrics stuff isn't ready so I couldn't fully confirm

File: gradle/wrapper/gradle-wrapper.properties (hunk 6 above)
Comment (Cervator) [reply gradle/wrapper/gradle-wrapper.properties:6]: Yep it did that for me. Left it in for my commit.
